src/collapse/perf.rs

```diff
@@ -390,87 +357,84 @@ impl PerfState {
-                Error::Parse(format!(
-                    "unable to parse file: {}, caused by: {}",
-                    module, e
-                ))
-            })?;
-            entry.insert(addr2line::Context::new(file)?);
+            let file =
+                File::open(module).map_err(|_| format!("unable to open file: {}", module))?;
+            // Using memmap is unsafe because the underlying file may be modified,
+            // which would cause undefined behavior. It's unlikely to be a problem in practice
+            // since we shouldn't run very long. The potential speed-up is worth it.
+            let map = unsafe {
+                memmap::Mmap::map(&file)
+                    .map_err(|_| format!("unable to memmap file: {}", module))?
+            };
+            let file = &object::File::parse(&*map)
+                .map_err(|_| format!("unable to parse file: {}", module))?;
+            entry.insert(addr2line::Context::new(file).map_err(|e| e.to_string())?);
         };
         let ctx = &self.addr2line_contexts[module];
 
-        let mut frames = ctx.find_frames(addr).map_err(|e| {
-            Error::Parse(format!(
-                "unable to parse frames from module: {}, at address: {}. caused by: {}",
-                module, addr, e
-            ))
+        let mut frames = ctx.find_frames(addr).map_err(|_| {
+            format!(
+                "unable to parse frames from module: {}, at address: {}",
+                module, addr
+            )
         })?;
 
-        let mut full_func = SmallVec::<[String; 1]>::new();
-        let mut total_length = 0;
+        let mut funcs = SmallVec::<[String; 1]>::new();
         while let Some(frame) = frames.next().unwrap() {
-            let mut function = frame
+            let func = frame
                 .function
                 .map(|func| {
-                    let name = remove_discriminator(
-                        func.raw_name()
-                            .map(Cow::from)
-                            .unwrap_or_else(|_| Cow::from("??")),
-                    );
-                    strip_rust_hash(addr2line::demangle_auto(name, func.language).as_ref())
-                        .to_string()
+                    let name = func.raw_name().map(Cow::from);
+                    let name = name.unwrap_or_else(|_| Cow::from("??"));
+                    let name = remove_discriminator(name);
+                    let name = match func.language {
+                        Some(gimli::DW_LANG_Rust) => {
+                            // Using rustc_demangle directly here since addr2line::demangle_auto
+                            // doesn't remove trailing hash value from Rust symbol names.
+                            // See https://github.com/gimli-rs/addr2line/issues/108
+                            format!("{:#}", rustc_demangle::demangle(&name))
+                        }
+                        lang => addr2line::demangle_auto(name, lang).to_string(),
```

Comment:
Can we not keep using `Cow<str>` here?

Reply:
Can we do that when we need to format for the Rust case?

Reply:
Yup -- just means that that will be a `Cow::Owned`!

Reply:
When I try this, the compiler complains. What can I do here?
```Rust
                .map(|func| {
                    let name = func.raw_name().map(Cow::from);
                    let name = name.unwrap_or_else(|_| Cow::from("??"));
                    let name = remove_discriminator(name);
                    match func.language {
                        Some(gimli::DW_LANG_Rust) => {
                            // Using rustc_demangle directly here since addr2line::demangle_auto
                            // doesn't remove trailing hash value from Rust symbol names.
                            // See https://github.com/gimli-rs/addr2line/issues/108
                            // Also the "{:#}" formatting is required to not include the
                            // trailing hash.
                            Cow::Owned(format!("{:#}", rustc_demangle::demangle(&name)))
                        }
                        lang => addr2line::demangle_auto(name, lang),
                    }
                })
                .unwrap_or_else(|| Cow::from("??"));
```

```
error[E0515]: cannot return value referencing function parameter `func`
   --> src/collapse/perf.rs:429:33
    |
417 |                     let name = func.raw_name().map(Cow::from);
    |                                ---- `func` is borrowed here
...
429 |                         lang => addr2line::demangle_auto(name, lang),
    |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returns a value referencing data owned by the current function
```

Reply:
I think it'll be easier to first simplify the code now that https://github.com/gimli-rs/addr2line/issues/108 has landed, and then try to figure it out?

Reply:
Don't we need to wait for a new release of addr2line on crates.io before we can simplify that?

Reply:
I pushed a commit that points Cargo.toml at the revision of addr2line with the fix we need. Are you ok with that, or should we just wait for a new release?

Also, let me know if you're satisfied with how Cow is being used now.